pkg/receive/head_series_limiter.go

```diff
@@ -0,0 +1,189 @@
+// Copyright (c) The Thanos Authors.
+// Licensed under the Apache License 2.0.
+
+package receive
+
+import (
+	"context"
+	"net/http"
+	"net/url"
+	"sync"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+	"github.com/thanos-io/thanos/pkg/errors"
+	"github.com/thanos-io/thanos/pkg/httpconfig"
+	"github.com/thanos-io/thanos/pkg/promclient"
+)
+
+// headSeriesLimiter encompasses active series limiting logic.
+type headSeriesLimiter interface {
+	QueryMetaMonitoring(context.Context, log.Logger) error
+	isUnderLimit(string, log.Logger) (bool, error)
+}
+
+// headSeriesLimit implements activeSeriesLimiter interface.
+type headSeriesLimit struct {
+	mtx                    sync.RWMutex
+	limit                  map[string]uint64
+	tenantCurrentSeriesMap map[string]float64
+	defaultLimit           uint64
+
+	metaMonitoringURL    *url.URL
+	metaMonitoringClient *http.Client
+	metaMonitoringQuery  string
+
+	configuredTenantLimit  *prometheus.GaugeVec
+	configuredDefaultLimit prometheus.Gauge
+	limitedRequests        *prometheus.CounterVec
+	metaMonitoringErr      prometheus.Counter
+}
+
+func NewHeadSeriesLimit(w WriteLimitsConfig, registerer prometheus.Registerer, logger log.Logger) *headSeriesLimit {
+	limit := &headSeriesLimit{
+		metaMonitoringURL:   w.GlobalLimits.metaMonitoringURL,
+		metaMonitoringQuery: w.GlobalLimits.MetaMonitoringLimitQuery,
+		defaultLimit:        w.DefaultLimits.HeadSeriesLimit,
+		configuredTenantLimit: promauto.With(registerer).NewGaugeVec(
+			prometheus.GaugeOpts{
+				Name: "thanos_receive_tenant_head_series_limit",
```

Comment:
I think you can skip the `tenant` in the name of this metric, given that it already has a tenant label.